### PR TITLE
feat: geo location enrichment in destinations

### PR DESCRIPTION
**Pull request by sanpj2292**
## Description of the change

We are looking to enrich the event with geo-location information that is available in `context.geo` object whenever customer enables a configuration on source.

### Destinations:
1. Freshsales
2. Facebook
3. Facebook_pixel
4. Facebook_offline_conversions
5. Facebook_conversions
6. google_adwords_enhanced_conversions
7. google_adwords_offline_conversions

Resolves INT-1096

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json [hunk 1]
```diff
@@ -86,15 +86,15 @@
...
+    "sourceKeys": ["traits.address.city", "context.traits.address.city", "context.geo.city"],
     "required": false,
     "metadata": {
       "type": ["toLower", "hashToSha256"]
     }
   },
   {
     "destKey": "st",
-    "sourceKeys": ["traits.address.state", "context.traits.address.state"],
+    "sourceKeys": ["traits.address.state", "context.traits.address.state", "context.geo.region"],
```
Comment (sanpj2292) [inline src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:97]: @krishna2020 @sandeepdsvs @utsabc 

Should we do this mapping ?

Example payload in context
```json
"geo": {
      "city": "Gurugram",
      "country": "IN",
      "ip": "223.190.82.63",
      "location": "28.459700,77.028200",
      "postal": "122001",
      "region": "Haryana",
      "timezone": "Asia/Kolkata"
    },
    "ip": "223.190.82.63"
```

**Comment by sanpj2292** [comment]
@krishna2020 @sandeepdsvs @utsabc

Should we do this mapping ?

Example geo-location payload in `context`
```json
"geo": {
    "city": "Gurugram",
    "country": "IN",
    "ip": "223.190.82.63",
    "location": "28.459700,77.028200",
    "postal": "122001",
    "region": "Haryana",
    "timezone": "Asia/Kolkata"
  },
  "ip": "223.190.82.63"
```

**Comment by sanpj2292** [comment]
One more thing to consider is, in some destinations we have `country` but it seems like `context.geo.country` is more close to countryCode. We can re-consider country -> context.geo.country mapping, if we see foresee breaking change.

**Review by koladilip — APPROVED** [review]

**Comment by koladilip** [comment]
> One more thing to consider is, in some destinations we have `country` but it seems like `context.geo.country` is more close to countryCode. We can re-consider country -> context.geo.country mapping, if we see foresee breaking change.

in such cases better to add to country code mapping (or create if not exists)

File: src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json
Comment (krishna2020) [inline src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:1]: Is this a sample one within facebook_conversion destination ? we do have other mapping configs in facebook_conversion

Comment (krishna2020) [inline src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:1]: will we have cases where an event for a destination will not have a mapping config ? If yes, what would be the plan to cover those cases ? Introduce mapping config for geo attributes requirement ?

Comment (sanpj2292) [reply src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:1]: No, this is full-change for `facebook_conversions` destination. In other mappings we do not have geo-location related fields present.

Comment (sanpj2292) [reply src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:1]: Except `FBCAppEventsConfig.json` where I did miss the change needed for `timezone` mapping

Comment (krishna2020) [reply src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:1]: Each destination had only one file changed, so was not sure whether all mapping configs got covered

**Comment by krishna2020** [comment]
> One more thing to consider is, in some destinations we have `country` but it seems like `context.geo.country` is more close to countryCode. We can re-consider country -> context.geo.country mapping, if we see foresee breaking change.

countryCode & country are considered as two different values. Can we check whether SDK's send country or countryCode ?
And we need to ask geo-location service to send both the values to transformer

**Review by koladilip — APPROVED** [review]

File: src/v0/destinations/google_adwords_enhanced_conversions/data/trackConfig.json [hunk 2]
```diff
@@ -107,7 +107,8 @@
       "context.traits.countryCode",
```
Comment (krishna2020) [inline src/v0/destinations/google_adwords_enhanced_conversions/data/trackConfig.json:107]: seems like mix of country/countryCode is being used

**Comment by sanpj2292** [comment]
In `context.geo` object we are getting country-code information as part of `country` field (as shown in the [example](https://github.com/rudderlabs/rudder-transformer/pull/2855#issuecomment-1831469712))

File: src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json [hunk 3]
```diff
@@ -115,7 +116,11 @@
   },
   {
     "destKey": "country",
-    "sourceKeys": ["traits.address.country", "context.traits.address.country"],
+    "sourceKeys": [
+      "traits.address.country",
+      "context.traits.address.country",
+      "context.geo.country"
```
[the remark this answers is not in the document]
Comment (sandeepdsvs) [reply src/v0/destinations/facebook_conversions/data/FBCUserDataConfig.json:122]: Freshsales allowes any string as a country. In this dev [doc](https://developer.freshsales.io/api/#list_all_contact_activities), country is sent as USA in some examples and as United States in some other

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

**Pull request closed without merging**